File: Data Structures & Algorithms/max-water-container/submission-0.go
func maxArea(heights []int) int {
	l := 0
	r := len(heights)-1
	area := 0
	for l<r {

		if heights[l] == min(heights[l], heights[r]) {
			area = max(area, heights[l]*(r-l))
			l++
		} else {
			area = max(area, heights[r]*(r-l))
			r--
		}
	}
	return area;
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Deque<Integer> stack = new ArrayDeque<>();
        int maxArea = 0;
        for(int i=0;i<=heights.length; i++){
            int h = i==heights.length? 0 : heights[i];
            while(!stack.isEmpty()&& h<heights[stack.peek()]){
                int r= stack.pop();
                int l = stack.isEmpty()?-1:stack.peek();
                maxArea = Math.max(heights[r]*(i-l-1),maxArea);
               
            }
            stack.push(i);
        }
        return maxArea;
    }
}
